> Getting an error during conversion of https://huggingface.co/tiiuae/falcon-mamba-7b-instruct, will investigate
> 
> ```
> Traceback (most recent call last):
>   File "/.../convert_hf_to_gguf.py", line 4078, in <module>
>     main()
>   File "/.../convert_hf_to_gguf.py", line 4072, in main
>     model_instance.write()
>   File "/.../convert_hf_to_gguf.py", line 390, in write
>     self.gguf_writer.write_kv_data_to_file()
>   File "/.../gguf-py/gguf/gguf_writer.py", line 240, in write_kv_data_to_file
>     kv_bytes += self._pack_val(val.value, val.type, add_vtype=True)
>                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/.../gguf-py/gguf/gguf_writer.py", line 855, in _pack_val
>     kv_data += encoded_val
> TypeError: can't concat list to bytearray
> ```
> 
> This is likely related with metadata extraction from the model card (when the `license` field is a list). @mofosyne is this already handled in one of your recent PRs?

Yeah I thought I already tackled this issue in https://github.com/ggerganov/llama.cpp/pull/8774 . Double checked by copying your falcon mamba model card metadata to my test repo and running against it. No issues detected.

Basically added a 'zero length array' check to `add_array()` to not add the kv in if that's the case.

--- a/convert_hf_to_gguf.py
+++ b/convert_hf_to_gguf.py
@@ -295,6 +295,7 @@ class Model:
                             gguf.MODEL_TENSOR.FFN_GATE_INP,
                             gguf.MODEL_TENSOR.POS_EMBD,
                             gguf.MODEL_TENSOR.TOKEN_TYPES,
+                            gguf.MODEL_TENSOR.SSM_CONV1D,
                         )
                     )
                     or not name.endswith(".weight")
@@ -2785,23 +2786,6 @@ class MambaModel(Model):
             self._tok_embd = data_torch
 
         return [(new_name, data_torch)]
-
-    def tensor_force_quant(self, name: str, new_name: str, bid: int | None, n_dims: int) -> gguf.GGMLQuantizationType | bool:
-        if bid is not None and new_name in (
-            self.format_tensor_name(
-                n, bid, ".weight" if name.endswith(".weight") else ""
-            )
-            for n in [
-                gguf.MODEL_TENSOR.SSM_CONV1D,
-                gguf.MODEL_TENSOR.SSM_X,
-                gguf.MODEL_TENSOR.SSM_DT,
-                gguf.MODEL_TENSOR.SSM_A,
-                gguf.MODEL_TENSOR.SSM_D,
-            ]
-        ):
-            return gguf.GGMLQuantizationType.F32
-
-        return super().tensor_force_quant(name, new_name, bid, n_dims)
 
 
 @Model.register("CohereForCausalLM")

--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -16122,6 +16122,9 @@ static ggml_type llama_tensor_get_type(quantize_state_internal & qs, ggml_type n
             case GGML_TYPE_Q6_K:   new_type = GGML_TYPE_Q8_0;   break;
             default: throw std::runtime_error("\nUnsupported tensor size encountered\n");
         }
+        if (tensor->ne[0] % ggml_blck_size(new_type) != 0) {
+            new_type = GGML_TYPE_F16;
+        }
         LLAMA_LOG_WARN(" - using fallback quantization %s\n", ggml_type_name(new_type));
         ++qs.n_fallback;
     }
@@ -16450,8 +16453,6 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
         // do not quantize Mamba's small yet 2D weights
         // NOTE: can't use LLM_TN here because the layer number is not known
         quantize &= name.find("ssm_conv1d.weight") == std::string::npos;
-        quantize &= name.find("ssm_x.weight")      == std::string::npos;
-        quantize &= name.find("ssm_dt.weight")     == std::string::npos;
 
         // do not quantize relative position bias (T5)
         quantize &= name.find("attn_rel_b.weight") == std::string::npos;

